### ISSUE TITLE
SAS token generated by azurerm_storage_account_sas is not working (current latest is 3.12)

### ISSUE THREAD

**Issue by DanielSSilva** [hashicorp/terraform-provider-azurerm#17558]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v1.1.9

### AzureRM Provider Version

3.12.0

### Affected Resource(s)/Data Source(s)

storage_account_sas

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "2.65"
    }
  }
#    backend "azurerm" {
#     use_azuread_auth = true
#   }

}

provider "azurerm" {
  features {}
  subscription_id = "XXXXX"
  tenant_id       = "XXXXX"
}

data "azurerm_key_vault" "terraform" {
  name                = var.key_vault.name
  resource_group_name = var.key_vault.resource_group_name
}

resource "azurerm_resource_group" "terraform" {
  name      = "rg-tempdebugstacc-dev"
  location  = "West Europe"
  tags      = var.tags
}

resource "azurerm_storage_account" "terraform" {
  name                            = "stdebugstacc1231"
  resource_group_name             = azurerm_resource_group.terraform.name
  location                        = azurerm_resource_group.terraform.location
  account_replication_type        = "LRS"
  account_tier                    = "Standard"
  tags                            = var.tags
  min_tls_version = "TLS1_2"
  #allow_nested_items_to_be_public = false
}

resource "azurerm_storage_container" "terraform" {
  name                  = "container"
  storage_account_name  = azurerm_storage_account.terraform.name
  container_access_type = "private"
}

data "azurerm_storage_account_sas" "terraform" {
  connection_string = azurerm_storage_account.terraform.primary_connection_string
  https_only        = true
  #signed_version    = "2021-06-08"
  start  = "2022-01-01"
  expiry = "2022-12-31"

  resource_types {
    service   = true
    container = true
    object    = true
  }

  services {
    blob  = true
    queue = true
    table = false
    file  = true
  }

  permissions {
    read    = true
    write   = true
    delete  = true
    list    = true
    add     = true
    create  = true
    update  = true
    process = true
    # tag     = true
    # filter  = true
  }
}

resource "azurerm_key_vault_secret" "terraform" {
  name         = "debug-storageaccount-sas-token"
  value        = data.azurerm_storage_account_sas.terraform.sas
  key_vault_id = data.azurerm_key_vault.terraform.id
}
```

### Debug Output/Panic Output

```shell
n/a
```

### Expected Behaviour

Using azcopy with the generated SAS token should work when using it to access the storage account

### Actual Behaviour

azcopy throws a 403 exception `AuthenticationErrorDetail: Signature did not match.`

### Steps to Reproduce

After successfully generating the SAS token through terraform with exactly the same permissions as the one generated on the portal, the signature (`sig`) is different, which then causes the calls made with the terraform generated SAS token to fail, whereas the ones made with the portal one works.

The terraform script I've shared is using version 2.85, which is the last version that I've used and was able to successfully generate & use the token.

Steps to reproduce:
1. Run the shared terraform script
2. Try to upload a file to "https://<storageaccountname>.blob.core.windows.net/container?<generated-sas-token>" -> it should succeed
3. changed the azurerm version to 3.12.0 (I've also tried with 3.2.0 and still had the error)
4. uncomment allow_nested_items_to_be_public, tag and filter permissions (required on this version)
5. run terraform again and try to upload a file to the same endpoint as before, with the new SAS token -> it fails

**Comment by DanielSSilva** [comment]
I'll leave my guesses/analysis here.
I've generated the token twice through the portal, with the exact same permissions and the output was the same, so I'm assuming that the HMAC-SHA256 algorithm is deterministic, which means that the terraform should output the same.

SAS token generation requires a signature, which is the
> the string-to-sign from the fields that compose the request, and then encode the string as UTF-8 and compute the signature by using the HMAC-SHA256 algorithm.
They also mentioned that 
> Version 2020-12-06 adds support for the signed encryption scope field.

When you use azcopy and you get the signature error message, it outputs what was used to sign, like
```
AuthenticationErrorDetail: Signature did not match. String to sign used was stdebugstacc1231
rwdlacuptf
bqtf
sco
2022-01-01
2022-12-31

https
2021-06-08


```
So, I've forced the sas token that's working to fail, by removing the "`f`" permission from the string, and then running azcopy and the output was:

```
AuthenticationErrorDetail: Signature did not match. String to sign used was stdebugstacc1231
rwdlacupt
bqtf
sco
2022-01-01
2022-12-31

https
2021-06-08


```
(the same without the `f`, as expected). Which leaves me to believe that all the configurable parameters are correctly specified, but there's one thing missing: the `signedEncryptionScope`.

I've also tried with the default sign version (2017-07-29) but still get the unauthorized error.

# My wild guess:
Could it be that, since the string to sign introduces the `signedEncryptionScope` and it's not being taken into consideration by the [terraform code](https://github.com/hashicorp/terraform-provider-azurerm/blob/ba4451a385c9e3690aecece0b7aa9691e836ed86/vendor/github.com/hashicorp/go-azure-helpers/storage/sas_token.go#L35-L43) (if I'm looking at the correct place), that's the reason why the signature is failing?

What doesn't make sense to me in this guess is that the 2.85 was released 8 months ago, and I'm assuming that the signedEncryptionScope was introduced around 2020-12-06

**Comment by magodo** [comment]
@DanielSSilva Thank you for this! Your guess is almost correct, see: https://docs.microsoft.com/en-us/rest/api/storageservices/create-account-sas#construct-the-signature-string.

As a workaround, you can set `tag` and `filter` to `false`, and don't set the `signed_version`, the generated SAS should still be able to use. Meanwhile, I've submit a PR to add on the missing `signedEncryptionScope`: https://github.com/hashicorp/go-azure-helpers/pull/123.

**Comment by DanielSSilva** [comment]
@magodo The workaround works and makes total sense since in the previous version there was no filter or tag, so setting them to false makes them not affect the signature. 
Out of curiosity, why is my guess "almost" correct and not totally correct? 😄 

Thank you for the quick reply and the workaround!

**Comment by magodo** [comment]
@DanielSSilva Sorry, I just meant that there are two factors result into the invalidation of SAS:

- The `signedEncryptionScope` that introduced since sign version 2020-12-06, which changes the format of the stuff to be signed
- The `tag` and `filter` (I don't know in which version they are introduced), which are apparently not supported in the default sign version, i.e. 2017-07-29

**Comment by DanielSSilva** [comment]
No worries! 😄  The `tag` and `filter` part totally make sense and I completely missed that part. Again, thanks for solving it so fast :)

**Comment by daniel-anova** [comment]
If i'm understanding correctly SaS with `filter` and `tag` set will always be invalid until this is fixed?

**Comment by magodo** [comment]
Since https://github.com/hashicorp/terraform-provider-azurerm/pull/21725 is merged that upgrades the go-azure-helpers to v0.56.0, so I'm gonna close this issue as the fix is in. Please feel free to reopen it if the issue remains.
